scan-command/src/main/java/io/ballerina/scan/internal/ProjectAnalyzer.java

```diff
@@ -69,68 +117,160 @@ private Consumer<DocumentId> analyzeDocument(Module module, ScannerContextImpl s
         }
+
         Module defaultModule = project.currentPackage().getDefaultModule();
-        List<String> defaultModuleFiles = defaultModule.documentIds()
-                .stream()
-                .map(documentId -> defaultModule.document(documentId).name())
-                .toList();
-        String documentName;
-        do {
-            documentName = String.format("%s-%s.bal", IMPORT_GENERATOR_FILE, UUID.randomUUID());
-        } while (defaultModuleFiles.contains(documentName));
         ModuleId defaultModuleId = defaultModule.moduleId();
-        DocumentId documentId = DocumentId.create(documentName, defaultModuleId);
-        DocumentConfig documentConfig = DocumentConfig.from(documentId, newImports.toString(), documentName);
+        DocumentId documentId = DocumentId.create(pluginImportsDocumentName, defaultModuleId);
+        DocumentConfig documentConfig = DocumentConfig.from(documentId, newImports.toString(),
+                pluginImportsDocumentName);
         defaultModule.modify().addDocument(documentConfig).apply();
-
         project.currentPackage().ballerinaToml().ifPresent(ballerinaToml -> {
             String tomlFileContent = ballerinaToml.tomlDocument().textDocument().toString();
             ballerinaToml.modify().withContent(tomlFileContent + tomlDependencies).apply();
         });
 
-        List<ScannerContext> scannerContextList = new ArrayList<>();
+        PackageResolution packageResolution = project.currentPackage().getResolution();
+        ResolvedPackageDependency rootPkgNode = new ResolvedPackageDependency(project.currentPackage(),
+                PackageDependencyScope.DEFAULT);
+        List<Package> directDependencies = packageResolution.dependencyGraph().getDirectDependencies(rootPkgNode)
+                .stream().map(ResolvedPackageDependency::packageInstance).toList();
+        Map<String, List<Rule>> externalAnalyzers = new HashMap<>();
+        for (Package pkgDependency : directDependencies) {
+            PackageManifest pkgManifest = pkgDependency.manifest();
+            String org = pkgManifest.org().value();
+            String name = pkgManifest.name().value();
+            String pluginName = org + FORWARD_SLASH + name;
+            if (pkgManifest.compilerPluginDescriptor().isEmpty() || !analyzerDescriptors.contains(pluginName)) {
+                continue;
+            }
 
-        // TODO: Implement the logic to pass scanner contexts to compiler plugins and perform external static analysis.
+            CompilerPluginDescriptor pluginDesc = pkgManifest.compilerPluginDescriptor().get();
+            List<URL> jarUrls = pluginDesc.dependencies().stream().map(dependency -> {
+                        try {
+                            return Path.of(dependency.getPath()).toUri().toURL();
+                        } catch (MalformedURLException ex) {
+                            throw new RuntimeException(ex);
```

Comment:
Better to use a custom message so that we know what exactly failed where. Also, we shouldn't be exposing Java exceptions if we can give a better and at least equally descriptive error.

Reply:
Can't this be ScanToolException too? We can uniformly handle `ScanToolException`s and consider everything else as unexpected failures then.